### PR TITLE
Add support for GitLab merge requests (MRs)

### PR DESCRIPTION
**Pull request by chamer81**
Adds an option to create a gitlab MR.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (ekohl) [inline .vscode/launch.json:1]: Please keep this out

File: bin/bundle [hunk 2]
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env ruby
```
Comment (ekohl) [inline bin/bundle:1]: Please keep this out

File: bin/rspec [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env ruby
```
Comment (ekohl) [inline bin/rspec:1]: Please keep this out

File: .rubocop_todo.yml [hunk 4]
```diff
@@ -1,70 +0,0 @@
-# This configuration was generated by
```
Comment (ekohl) [inline .rubocop_todo.yml:1]: Could you submit this as a PR first? It makes reviewing this PR easier.

File: lib/modulesync.rb [hunk 5]
```diff
@@ -1,4 +1,5 @@
 require 'fileutils'
+require 'gitlab'
```
Comment (ekohl) [inline lib/modulesync.rb:2]: If this becomes a hard dependency then it needs to be in the gemspec

**Comment by bittner** [comment]
Thank you @ekohl for the quick feedback. We'll take that immediately into account. We're still at fixing some rough edges of the implementation and doing internal review.

**Comment by bittner** [comment]
Relates to https://github.com/voxpupuli/modulesync/issues/80#issuecomment-391329471.

**Comment by bittner** [comment]
We have cleaned up the commits and I've consolidated the changes into a single commit.

There are still a few topics open: (at least from a first, quick glance)

- Apart from GitLab.com, we need to allow GitLab support for **private instances** ([`config.endpoint`](https://github.com/narkoz/gitlab#usage) is set for this); presumably, this needs to be configurable
- The added / updated **tests** don't seem complete
- The **lazy-loading** of the GitHub and GitLab dependencies is still pending realization

File: .gitignore [hunk 6]
```diff
@@ -1,5 +1,10 @@
...
 *.gem
 Gemfile.lock
 .bundle/
+.ruby-version
+.vscode/
+bin/bundle
+bin/rspec
+modules/
+tmp/
+vendor/
```
Comment (ekohl) [inline .gitignore:10]: This would typically be a separate commit.

File: Gemfile [hunk 7]
```diff
@@ -2,5 +2,7 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
 
 gemspec
 
+gem 'aruba'
```
Comment (ekohl) [inline Gemfile:5]: Why is this needed, looks redundant:
https://github.com/voxpupuli/modulesync/blob/7202e6bfc819a8e7323a03c9391cf45b2b6df9d2/modulesync.gemspec#L20

File: Gemfile [hunk 8]
```diff
@@ -2,5 +2,7 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
...
 gem 'cucumber', '< 3.0' if RUBY_VERSION < '2.1'
+gem 'gitlab' # gem 'gitlab', github: 'NARKOZ/gitlab'
 gem 'octokit', '~> 4.9'
```
Comment (ekohl) [inline Gemfile:8]: This was there, but I can't explain why because this should be sufficent:
https://github.com/voxpupuli/modulesync/blob/7202e6bfc819a8e7323a03c9391cf45b2b6df9d2/modulesync.gemspec#L27

File: Gemfile [hunk 9]
```diff
@@ -2,5 +2,7 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
 
 gemspec
 
+gem 'aruba'
 gem 'cucumber', '< 3.0' if RUBY_VERSION < '2.1'
+gem 'gitlab' # gem 'gitlab', github: 'NARKOZ/gitlab'
```
Comment (ekohl) [inline Gemfile:7]: This is redundant since it's in the gemspec

**Comment by mhutter** [comment]
Updated the first commit to clean up dependencies in `Gemfile` and gemspec. Added a second commit to refactor the existing code. This (as a side-effect) also implements lazy-loading and client reuse.

**Comment by mhutter** [comment]
Still looking into that failing cucumber test, not sure how that's related...

**Comment by mhutter** [comment]
One thing that caught my eye when moving code around: The Gitlab implementation will ensure the configured labels are added to the PR when the PR is created, but will NOT add any labels if there is already an open PR around.

The GitHub implementation will always update the existing PR - and fail horribly from looking at the code (note this was already broken before this PR)

@ekohl what is the intended behaviour?

**Comment by bittner** [comment]
@ekohl Can you clarify?

File: lib/modulesync.rb [hunk 10]
```diff
@@ -11,10 +10,7 @@
 require 'monkey_patches'
 
 GITHUB_TOKEN = ENV.fetch('GITHUB_TOKEN', '')
-
-Octokit.configure do |c|
-  c.api_endpoint = ENV.fetch('GITHUB_BASE_URL', 'https://api.github.com')
-end
+GITLAB_TOKEN = ENV.fetch('GITLAB_TOKEN', '')
```
Comment (ekohl) [inline lib/modulesync.rb:13]: I think we can avoid these constants and only use them inline since they're only used in a single place.

Comment (mhutter) [reply lib/modulesync.rb:13]: Good point, totally missed it. Will adjust accordingly

**Comment by ekohl** [comment]
This looks good to me. It's still marked as a draft so let me know when you think it's ready to be merged.

**Comment by mhutter** [comment]
I broke the check in my last commit, fixed.

I think the PR is ready, can you remove the "Draft" status @chamer81 ?

I'll open a new Issue to discuss the PR label behaviour. Fixing it should be trivial.

**Comment by ekohl** [comment]
Thanks!

**Comment by bittner** [comment]
Thanks everyone for the good collaboration! :1st_place_medal: :confetti_ball:

File: lib/modulesync.rb [hunk 11]
```diff
@@ -201,4 +168,20 @@ def self.update(options)
...
+
+  def self.create_pr_manager
+    github_token = ENV.fetch('GITHUB_TOKEN', '')
+    gitlab_token = ENV.fetch('GITLAB_TOKEN', '')
+
+    if !github_token.empty?
+      require 'modulesync/pr/github'
+      ModuleSync::PR::GitHub.new(github_token, ENV.fetch('GITHUB_BASE_URL', 'https://api.github.com'))
+    elsif !gitlab_token.empty?
+      require 'modulesync/pr/github'
```
Comment (raphink) [inline lib/modulesync.rb:180]: Should be `gitlab`

Comment (bittner) [reply lib/modulesync.rb:180]: Fixed via commit 2b5388230fb39dff85e3522db753d20e8c88dfab.
